cpp/util/ManagedArray.h

```diff
@@ -0,0 +1,146 @@
+    /*! \param size Size of the array to allocate.
+     */
+    ManagedArray(unsigned int size=0)
+    {
+        m_size = std::make_shared<unsigned int>(size);
+        // Hard code making an array of at least size 1.
+        m_data = std::shared_ptr<std::shared_ptr<T> >(
+            new std::shared_ptr<T>(new T[1], std::default_delete<T[]>()));
+        create_new_array(size);
+        reset();
+    }
+
+    //! Destructor (currently empty because data is managed by shared pointer).
+    ~ManagedArray() {}
+
+    //! Update size of the array.
+    /*! \param size New size of the array.
+     */
+    void resize(unsigned int size)
+    {
+        if (size != *m_size)
+        {
+            *m_size = size;
+            reallocate();
+        }
+    }
+
+    //! Allocate new memory for the array.
+    /*! This function primarily serves as a way to reallocate new memory from
+     *  the Python API if needed.
+     * \param size New size of the array.
+     */
+    void reallocate()
+    {
+        create_new_array(*m_size);
+        reset();
+    }
+
+    //! Reset the contents of array to be 0.
+    void reset()
+    {
+        if (*m_size != 0)
+        {
+            memset((void*) get(), 0, sizeof(T) * (*m_size));
+        }
+    }
+
+    //! Return the underlying pointer (requires two levels of indirection).
+    T *get() const
+    {
+        std::shared_ptr<T> * tmp = m_data.get();
+        return (*tmp).get();
+    }
+
+    //! Writeable index into array.
+    T &operator[](unsigned int index)
+    {
+        if (index >= *m_size)
+        {
+            throw std::runtime_error("Attempted to access data out of bounds.");
+        }
+        return get()[index];
+    }
+
+    //! Read-only index into array.
+    const T &operator[](unsigned int index) const
+    {
+        if (index >= *m_size)
+        {
+            throw std::runtime_error("Attempted to access data out of bounds.");
+        }
+        return get()[index];
+    }
+
+    //! Get the size of the current array.
+    unsigned int size() const
+    {
+        return *m_size;
+    }
+
+    //! Make a deep copy of this array (i.e. copy the underlying data).
+    /*! This function returns by value since all memory ownership information
+     *  is transmitted via shared pointers. 
+     */
+    void *deepCopy()
+    {
+        ManagedArray<T> *deep_copy = new ManagedArray<T>(size());
+        // This *m_data will be a different shared_ptr than the current
+        // instance's, so assigning to it directly shouldn't affect any other
+        // arrays pointing to the memory of the current one and reallocating
+        // the memory of others won't affect this one.
+        *((*deep_copy).m_data) = *m_data;
+        return deep_copy;
+    }
+        
+private:
+    //! Reallocate the data array into a new chunk of memory.
+    void create_new_array(unsigned int size)
+    {
+        // Always allocate at least size 1
+        unsigned int new_size = size > 1 ? size : 1;
```

Comment:
Can you help me understand why we potentially change the size from 0?

Reply:
Ah this is a holdover from when I couldn't remember whether the C++ standard permitted 0 length arrays. I didn't want to write non-portable code and this was my kludgy solution. But I believe I looked at the standard afterward and found that 0-length _heap_ arrays are allowed post C++0x, but the only valid (and required) operation on it is deletion (dereferencing is UB). Correct me if I'm wrong.

Reply:
Yeah, I think you can create them, but you must also delete them.